orttraining/orttraining/training_ops/cuda/tensor/pad_by_axis_impl.cu

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cuda/tensor/pad_by_axis_impl.h"
+#include "core/providers/cuda/cu_inc/common.cuh"
+
+namespace onnxruntime {
+namespace cuda {
+
+constexpr int kBlockSize = 256;
+constexpr int kNumUnroll = 4;
+
+template <typename T>
+__global__ void FillOutputWithIndexKernel(const CUDA_LONG N,
+                                          const fast_divmod output_element_stride_fdm,
+                                          const T* input_data,
+                                          const int64_t* indices_data,
+                                          T* output_data) {
+  CUDA_LONG idx = blockDim.x * blockIdx.x + threadIdx.x;
+  CUDA_LONG id = idx * kNumUnroll;
+
+  T input[kNumUnroll];
+  if (id < N) {
+#pragma unroll
+    for (int i = 0; i < kNumUnroll; ++i) {
+      CUDA_LONG li = id + i;
+      if (li < N) {
+        input[i] = input_data[li];
+      }
+    }
+  }
+
+#pragma unroll
+  for (int i = 0; i < kNumUnroll; ++i) {
+    CUDA_LONG li = id + i;
+    if (li < N) {
+      int row_index, col_index;
+      output_element_stride_fdm.divmod(li, row_index, col_index);
+      output_data[indices_data[row_index] * output_element_stride_fdm.d_ + col_index] = input[i];
```

Comment:
Is it necessary to check the index not lager than the output size ?

Reply:
If the indices_data[row_index] is valid, then the output index should be valid. But you are right, let me add an assert for indices_data[row_index] in case it is passed in an invalid value.